drafts/inlinemacros.md

```diff
@@ -165,99 +170,66 @@ positions:
-`inline` are handled analogously. For instance, a type like
-
-    inline (inline Int, Boolean) => String
-
-would be seen inside a macro as
-
-    (Int, Expr[Boolean]) => Expr[String]
-
-A dual transformation is applied to a macro's result. Normal types are prefixed with `inline` until
-a type is of the form `Expr[T]` in which case it is replaced by `T`. So if a macro body returned
-a list of type
-
-    List[(String, Expr[T])]
-
-the type of the macro itself would be
-
-    inline List[inline (inline String, T)]
-
-## Whitebox vs Black Box
-
-Normaly, we would require that an inline method has an explicit result type.
-This restriction is in place so that we can expand inline functions and macros
-in a separate phase after type checking.
-
-We could allow inline functions with missing result type (so called "whitebox macros"),
-maybe controlled by a language import such as
-
-    import language.whiteboxMacros
-
+where `{ ... }` is some block of Scala code, called macro body. (In fact, `macro` may prefix arbitrary expressions,
+but blocks are used most commonly). 
+
+In a macro expression, `macro` stands for a function 
+`def macro[T](body: implicit scala.meta.macros.Context => scala.meta.Term[T]): T = ???` declared in `scala.Predef`.
+Description of the functionality exposed by the `scala.meta` metaprogramming library 
+is outside of the scope of this proposal.
+
+Macro expressions can appear both in the bodies of inline methods
+(then their expansion is going to be deferred until these methods expand) and in normal code
+(in that case, expansion will take place immediately at the place where the macro expression is written).
+
+Macro bodies can only reference the following names in their environment:
+
+ 1. type parameters of enclosing inline methods and classes/traits containing an enclosing inline method as a member,
+ 2. terms parameters of enclosing inline methods (where the parameter may, or may not be, marked inline),
+ 3. `this` references to classes/traits containing an enclosing inline method as a member,
+ 4. inline values,
+ 5. inline methods,
+ 6. anything that's global (i.e. a class/trait/object without an outer reference).
+
+Names referenced in macro bodies undergo certain changes. References to inline and global definitions are left untouched,
+but local terms of type `T` are transformed into terms of type `scala.meta.Term[T]`, 
+and local types `U` are transformed into terms of type `scala.meta.Type`. In other words, 
+definitions that are statically available outside macro bodies remain available in macro bodies,
+whereas term and type parameters of enclosing inline methods become available as their representations.
+
+## Macro expansion
+
+During typechecking, the compiler treats macro expressions, i.e. invocations of `Predef.macro`,
+as normal method calls, typechecking macro bodies and performing type inference if necessary, but nothing else.
```

Comment:
We don't need to prohibit eta expansion of `Predef.macro`, because I can't see how people could make it typecheck :)